### ISSUE TITLE
stack overflow by enum and const

### ISSUE THREAD

**Issue by mikhail-m1** [rust-lang/rust#36163]
while I was working on #35309 I found example not detected by static recursion checker

```
const A : i32 = Foo::B;
enum Foo { B = A }
fn main() {}
```

I'm going to fix it
